core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/AwsChunkedV4PayloadSigner.java

```diff
@@ -0,0 +1,167 @@
+                .firstMatchingHeader(Header.CONTENT_LENGTH)
+                .orElseThrow(() -> new IllegalArgumentException(Header.CONTENT_LENGTH + " must be specified!"));
+
+            request.putHeader(X_AMZ_DECODED_CONTENT_LENGTH, contentLength)
+                   .removeHeader(Header.CONTENT_LENGTH);
+        } else {
+            // decoded header is already there, so remove content-length just to be sure it's gone
+            request.removeHeader(Header.CONTENT_LENGTH);
+        }
+    }
+
+    @Override
+    public ContentStreamProvider sign(ContentStreamProvider payload, V4Context v4Context) {
+        SdkHttpRequest.Builder request = v4Context.getSignedRequest();
+        moveContentLength(request);
+
+        String checksum = request.firstMatchingHeader(X_AMZ_CONTENT_SHA256).orElseThrow(
+            () -> new IllegalArgumentException(X_AMZ_CONTENT_SHA256 + " must be set!")
+        );
+
+        InputStream inputStream = payload != null ? payload.newStream() : new StringInputStream("");
+        ChunkedEncodedInputStream.Builder chunkedEncodedInputStreamBuilder = ChunkedEncodedInputStream
+            .builder()
+            .inputStream(inputStream)
+            .chunkSize(chunkSize)
+            .header(chunk -> Integer.toHexString(chunk.length).getBytes(StandardCharsets.UTF_8));
+
+        switch (checksum) {
+            case STREAMING_SIGNED_PAYLOAD: {
+                RollingSigner rollingSigner = new RollingSigner(v4Context.getSigningKey(), v4Context.getSignature());
+                setupSigExt(chunkedEncodedInputStreamBuilder, rollingSigner);
+                break;
+            }
+            case STREAMING_UNSIGNED_PAYLOAD_TRAILER:
+                setupChecksumTrailer(chunkedEncodedInputStreamBuilder);
+                break;
+            case STREAMING_SIGNED_PAYLOAD_TRAILER: {
+                RollingSigner rollingSigner = new RollingSigner(v4Context.getSigningKey(), v4Context.getSignature());
+                setupSigExt(chunkedEncodedInputStreamBuilder, rollingSigner);
+                setupSigTrailer(chunkedEncodedInputStreamBuilder, rollingSigner);
+                setupChecksumTrailer(chunkedEncodedInputStreamBuilder);
+                break;
+            }
+            default:
+                throw new UnsupportedOperationException();
+        }
+
+        return chunkedEncodedInputStreamBuilder::build;
+    }
+
+    @Override
+    public Publisher<ByteBuffer> signAsync(Publisher<ByteBuffer> payload, V4Context v4Context) {
+        throw new UnsupportedOperationException();
+    }
+
+    private void setupSigExt(ChunkedEncodedInputStream.Builder builder, RollingSigner rollingSigner) {
+        builder.addExtension(
+            chunk -> Pair.of(
+                "chunk-signature".getBytes(StandardCharsets.UTF_8),
+                rollingSigner.sign(
+                    previousSignature ->
+                        "AWS4-HMAC-SHA256-PAYLOAD" + SignerConstant.LINE_SEPARATOR +
```

Comment:
Same here, let's create a variable. Can we use String.join instead of concatenation?